src/gen9_hevc_encoder.c

```diff
@@ -0,0 +1,7376 @@
+                          dri_bo *bo,
+                          unsigned int offset,
+                          unsigned int data)
+{
+    struct gpe_mi_conditional_batch_buffer_end_parameter mi_cond_end;
+
+    memset(&mi_cond_end, 0, sizeof(mi_cond_end));
+    mi_cond_end.offset = offset;
+    mi_cond_end.bo = bo;
+    mi_cond_end.compare_data = data;
+    mi_cond_end.compare_mask_mode_disabled = 0;
+    gen9_gpe_mi_conditional_batch_buffer_end(ctx, batch,
+                                             &mi_cond_end);
+}
+
+static VAStatus
+gen9_hevc_ensure_surface(VADriverContextP ctx,
+                         struct gen9_hevc_encoder_state *priv_state,
+                         struct object_surface *obj_surface,
+                         int reallocate_flag)
+{
+    VAStatus va_status = VA_STATUS_SUCCESS;
+    int update = 0;
+    unsigned int fourcc = VA_FOURCC_NV12;
+
+    if (!obj_surface) {
+        va_status = VA_STATUS_ERROR_INVALID_PARAMETER;
+
+        goto EXIT;
+    }
+
+    if ((priv_state->bit_depth_luma_minus8 > 0)
+        || (priv_state->bit_depth_chroma_minus8 > 0)) {
+        if (obj_surface->fourcc != VA_FOURCC_P010) {
+            update = 1;
+            fourcc = VA_FOURCC_P010;
+        }
+    }
+    else if (obj_surface->fourcc != VA_FOURCC_NV12) {
+        update = 1;
+        fourcc = VA_FOURCC_NV12;
+    }
+
+    /* (Re-)allocate the underlying surface buffer store, if necessary */
+    if (!obj_surface->bo || update) {
+        if (reallocate_flag) {
+            struct i965_driver_data * const i965 = i965_driver_data(ctx);
+
+            i965_destroy_surface_storage(obj_surface);
+
+            va_status = i965_check_alloc_surface_bo(ctx,
+                                                    obj_surface,
+                                                    i965->codec_info->has_tiled_surface,
+                                                    fourcc,
+                                                    SUBSAMPLE_YUV420);
+        }
+        else
+            va_status = VA_STATUS_ERROR_INVALID_PARAMETER;
+    }
+
+EXIT:
+    return va_status;
+}
+
+static void
+gen9_hevc_free_surface_private(void **data)
+{
+    struct gen9_hevc_surface_priv *surface_priv = NULL;
+    int i = 0;
+
+    surface_priv = (struct gen9_hevc_surface_priv *)(*data);
+    if (!surface_priv)
+        return;
+
+    for (i = 0;i < HEVC_SCALED_SURFS_NUM;i++) {
+      if (surface_priv->scaled_surface_obj[i]) {
+          i965_DestroySurfaces(surface_priv->ctx, &surface_priv->scaled_surface_id[i], 1);
+          surface_priv->scaled_surface_id[i] = VA_INVALID_SURFACE;
+          surface_priv->scaled_surface_obj[i] = NULL;
+      }
+    }
+
+    i965_DestroySurfaces(surface_priv->ctx, &surface_priv->surface_id_nv12, 1);
+    surface_priv->surface_id_nv12 = VA_INVALID_SURFACE;
+    surface_priv->surface_obj_nv12 = NULL;
+
+    dri_bo_unreference(surface_priv->motion_vector_temporal_bo);
+
+    free(surface_priv);
+}
+
```

Comment:
The current code doesn't consider the change of width/height for the encoded frame especially when the encoded frame is also sued when switching from higher resolution to lower resolution.